py-polars/tests/unit/operations/test_comparison.py

```diff
@@ -144,3 +152,197 @@ def test_missing_equality_on_bools() -> None:
         True,
         False,
     ]
+
+
+def isnan(x: Any) -> bool:
+    return isinstance(x, float) and math.isnan(x)
+
+
+def reference_ordering_propagating(lhs: Any, rhs: Any) -> str | None:
+    # normal < nan, nan == nan, nulls propagate
+    if lhs is None or rhs is None:
+        return None
+
+    if isnan(lhs) and isnan(rhs):
+        return "="
+
+    if isnan(lhs) or lhs > rhs:
+        return ">"
+
+    if isnan(rhs) or lhs < rhs:
+        return "<"
+
+    return "="
+
+
+def reference_ordering_missing(lhs: Any, rhs: Any) -> str:
+    # null < normal < nan, nan == nan, null == null
+    if lhs is None and rhs is None:
+        return "="
+
+    if lhs is None:
+        return "<"
+
+    if rhs is None:
+        return ">"
+
+    if isnan(lhs) and isnan(rhs):
+        return "="
+
+    if isnan(lhs) or lhs > rhs:
+        return ">"
+
+    if isnan(rhs) or lhs < rhs:
+        return "<"
+
+    return "="
+
+
+def verify_total_ordering(
+    lhs: Any, rhs: Any, dummy: Any, dtype: pl.PolarsDataType
+) -> None:
+    assert dummy is not None
+    ref = reference_ordering_propagating(lhs, rhs)
+    refmiss = reference_ordering_missing(lhs, rhs)
+
+    # Add dummy variable so we don't broadcast or do full-null optimization.
+    df = pl.DataFrame(
+        {"l": [lhs, dummy], "r": [rhs, dummy]}, schema={"l": dtype, "r": dtype}
+    )
+
+    ans = df.select(
+        (pl.col("l") == pl.col("r")).alias("eq"),
+        (pl.col("l") != pl.col("r")).alias("ne"),
+        (pl.col("l") < pl.col("r")).alias("lt"),
+        (pl.col("l") <= pl.col("r")).alias("le"),
+        (pl.col("l") > pl.col("r")).alias("gt"),
+        (pl.col("l") >= pl.col("r")).alias("ge"),
+        pl.col("l").eq_missing(pl.col("r")).alias("eq_missing"),
+        pl.col("l").ne_missing(pl.col("r")).alias("ne_missing"),
+    )
+
+    ans_correct_dict = {
+        "eq": [ref and ref == "="],  # "ref and X" propagates ref is None
+        "ne": [ref and ref != "="],
+        "lt": [ref and ref == "<"],
+        "le": [ref and (ref == "<" or ref == "=")],
+        "gt": [ref and ref == ">"],
+        "ge": [ref and (ref == ">" or ref == "=")],
+        "eq_missing": [refmiss == "="],
+        "ne_missing": [refmiss != "="],
+    }
+    ans_correct = pl.DataFrame(
+        ans_correct_dict, schema={c: pl.Boolean for c in ans_correct_dict}
+    )
+
+    assert_frame_equal(ans[:1], ans_correct)
+
+
+def verify_total_ordering_broadcast(
```

Comment:
These tests seem quite complete, though for my personal tastes I would say they are too complicated. If a test fails, someone will have to spend quite some time figuring out what these tests do exactly / where it fails. Then again, this is probably the best way to avoid duplication. So... :man_shrugging: I guess it's fine.

Reply:
I've actually caught multiple bugs due to the completeness of the test 😅 Testing every type and every combination of reasonably interesting inputs makes me confident it's correct.